### ISSUE TITLE
Infinite recursion loop when testing with MagicMock

### ISSUE THREAD

**Issue by M3t0r** [dateutil/dateutil#455]
```python
from dateutil.parser import parse
from unittest.mock import MagicMock

parse(MagicMock())
```
results in an infinite recursion loop.

This is very hindering if you want to test code with mocks that parses dates with dateutil but requires MagicMock because you code is iterating over the mock first.

Do you have a best practice for testing code that uses parse?

**Comment by jbrockmendel** [comment]
Not sure I understand.  What are you trying to mock?

**Comment by M3t0r** [comment]
For example:
```python
def tested_function(mocked_object):
    return parse(mocked_object.tags["timestamp]) + timedelta(hours=5)
```
`.tags` has to be a MagicMock to support the `__getitem__` call. Also I could prepare the `mocked_object` like so:
```python
mocked_object = Mock(name="object")
mocked_object.tags = {"timestamp": "1970-01-01T00:00Z"}
```
But then I'd have a concrete value, not a mock.

I think `dateutil.parser.parse` should detect mocks and return a new mock.

**Comment by pganssle** [comment]
@M3t0r I think anything that special-cases mocked objects sorta defeats the point, doesn't it? That would make any mocked tests invalid...

I still don't totally understand what the problem is, here - if you want `parser.parse` to do nothing, then you should just mock out `parser` so that it returns a mock...

**Comment by jbrockmendel** [comment]
> I think anything that special-cases mocked objects sorta defeats the point

I was thinking the same thing.

**Comment by pganssle** [comment]
Marking this wontfix for now, but I'll leave it open a bit longer in case @M3t0r has some clarifying example why this might be an actual issue for us.

**Comment by M3t0r** [comment]
Nah, you're probably right.
I had a collection of functions that do very different things with the same input and I just wanted to make sure they don't crash. So my simple solution was to just call all of them with a mock object as argument and be ok with them as long as they didn't throw an exception. But one function tries to parse a date string and that ended in an infinity loop in my test suite.
I guess I have to write more explicit tests for them. With `@patch` and all.

**Comment by bobwhitelock** [comment]
I recently ran into this issue myself locally, and I found it a bit tricky to debug. In my case this was with a large test suite and a slightly misconfigured local environment, leading to a couple of tests locally taking a different code path to what they normally would and unexpectedly passing a `MagicMock` to `parse`. From an external perspective this issue then manifested itself by the test process eventually using up all my machine's memory and getting killed by the OOM killer.

While I agree that special-casing mock objects seems to go against their purpose, and I can't see another way to distinguish these from other valid inputs, maybe something else could be done to make situations like this easier to debug. Would limiting the number of iterations of this loop work to make the issue a bit more obvious, maybe to 1000 or so (at least on my machine, this would have occurred before all my memory was used)? I don't know enough to know whether this many iterations is ever likely to occur with valid inputs which will ultimately end up being parsed as a date - depending on whether this is the case maybe just a warning could be given to look into the input, or if not an exception raised?

This would not seem to be special-casing mocks and would also prevent an infinite loop for other bad inputs, for instance a class like this will currently also cause an infinite loop:

```python
from dateutil.parser import parse

class Foo:
    def read(*args, **kwargs):
        return 'a'

parse(Foo())
```
